torch/_inductor/scheduler.py

```diff
@@ -2142,8 +2245,323 @@ def can_eliminate_user(user: NodeUser) -> bool:
+                    node.pred_buffers.append(sched_buf)
+                    sched_buf.succ_nodes.add(node)
+            for dep_name in {
+                dep.name for dep in node.read_writes.reads - node.unmet_dependencies
+            }:
+                input_buf = self.name_to_input_buf.get(dep_name)
+                if input_buf:
+                    node.pred_buffers.append(input_buf)
+                    input_buf.succ_nodes.add(node)
+
+    def estimate_peak_memory(self, nodes: List[BaseSchedulerNode]) -> int:
+        """
+        Given a list of nodes in their execution order, estimate the peak memory, by
+        keeping track of the liveliness of SchedulerBuffers and InputBuffers.
+        """
+
+        # map each scheduler buffer to its size, start time, and end time
+        @dataclasses.dataclass
+        class BufferInfo:
+            buffer: Union[SchedulerBuffer, InputBuffer]
+            size_alloc: int
+            size_free: int
+            start_time: int
+            end_time: int
+
+        name_to_buf_info: Dict[str, BufferInfo] = {}
+        node_name_to_time: Dict[str, int] = {}
+
+        # assign start_time
+        for buf_name, input_buf in self.name_to_input_buf.items():
+            name_to_buf_info[buf_name] = BufferInfo(
+                input_buf, input_buf.size_free, input_buf.size_free, 0, 0
+            )
+        for t, node in enumerate(nodes):
+            node_name_to_time[node.get_name()] = t
+            for sched_buf in node.get_outputs():
+                name_to_buf_info[sched_buf.get_name()] = BufferInfo(
+                    sched_buf, sched_buf.size_alloc, sched_buf.size_free, t, t
+                )
+
+        # assign end_time
+        for buf_name, buf_info in name_to_buf_info.items():
+            succ_node_time = [
+                node_name_to_time[succ_node.get_name()]
+                for succ_node in buf_info.buffer.succ_nodes
+                if succ_node.get_name() in node_name_to_time
+            ]
+            if succ_node_time:
+                buf_info.end_time = max(succ_node_time)
+
+        # the end time of output buffers should be at the end of the horizon
+        for buf_name in V.graph.get_output_names():
+            if buf_name in name_to_buf_info:
+                name_to_buf_info[buf_name].end_time = len(nodes) - 1
+
+        # incremental memory changes at each time period
+        memory = [0 for _ in range(len(nodes) + 1)]
+
+        # for each buffer, update memory when created and when freed
+        for buf_name, buf_info in name_to_buf_info.items():
+            memory[buf_info.start_time] += buf_info.size_alloc
+            memory[buf_info.end_time + 1] -= buf_info.size_free
+
+        # get peak memory by compute the cumulative memories
+        max_memory = 0
+        cur_memory = 0
+        for t in range(len(nodes) + 1):
+            cur_memory += memory[t]
+            max_memory = max(max_memory, cur_memory)
+
+        return max_memory
```

Comment:
curious: is there a way to also get a dictionary of "node_index -> cur_memory at that index" from this function above? (maybe we can return it as a dict?) I believe it will be useful for FSDP memory-based prefetching as well

Reply:
I think the `cur_memory` should give this information, let me return both max_memory and `cur_memory` for this function.

Reply:
The function now returns both peak memory and memories at each step.